pkg/joinserver/http_interop.go

```diff
@@ -16,17 +16,142 @@ package joinserver
 
 import (
 	"context"
+	"time"
 
+	"go.thethings.network/lorawan-stack/pkg/encoding/lorawan"
 	"go.thethings.network/lorawan-stack/pkg/errors"
 	"go.thethings.network/lorawan-stack/pkg/interop"
+	"go.thethings.network/lorawan-stack/pkg/log"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/types"
 )
 
-type interopServer struct {
-	JS *JoinServer
+type interopHandler interface {
+	HandleJoin(context.Context, *ttnpb.JoinRequest) (*ttnpb.JoinResponse, error)
+	GetHomeNetID(context.Context, types.EUI64, types.EUI64) (*types.NetID, error)
+	GetAppSKey(context.Context, *ttnpb.SessionKeyRequest) (*ttnpb.AppSKeyResponse, error)
 }
 
-var errInteropNotImplemented = errors.DefineUnimplemented("interop_not_implemented", "interop endpoint not implemented")
+type interopServer struct {
+	JS interopHandler
+}
 
 func (srv interopServer) JoinRequest(ctx context.Context, req *interop.JoinReq) (*interop.JoinAns, error) {
-	return nil, errInteropNotImplemented
+	ctx = log.NewContextWithField(ctx, "namespace", "joinserver/interop")
+
+	var cfList *ttnpb.CFList
+	if len(req.CFList) > 0 {
+		cfList = new(ttnpb.CFList)
+		if err := lorawan.UnmarshalCFList(req.CFList, cfList); err != nil {
+			return nil, interop.ErrMalformedMessage.WithCause(err)
+		}
+	}
+	var dlSettings ttnpb.DLSettings
+	if err := lorawan.UnmarshalDLSettings(req.DLSettings, &dlSettings); err != nil {
+		return nil, interop.ErrMalformedMessage.WithCause(err)
+	}
+
+	res, err := srv.JS.HandleJoin(ctx, &ttnpb.JoinRequest{
+		RawPayload:         req.PHYPayload,
+		DevAddr:            types.DevAddr(req.DevAddr),
+		SelectedMACVersion: ttnpb.MACVersion(req.MACVersion),
+		NetID:              types.NetID(req.SenderID),
+		DownlinkSettings:   dlSettings,
+		RxDelay:            req.RxDelay,
+		CFList:             cfList,
+	})
+	if err != nil {
+		switch {
+		case errors.Resemble(err, errDecodePayload),
+			errors.Resemble(err, errWrongPayloadType),
+			errors.Resemble(err, errNoDevEUI),
+			errors.Resemble(err, errNoJoinEUI):
+			return nil, interop.ErrMalformedMessage.WithCause(err)
+		case errors.Resemble(err, errAddressNotAuthorized):
+			return nil, interop.ErrActivation.WithCause(err)
+		case errors.Resemble(err, errMICMismatch):
+			return nil, interop.ErrMIC.WithCause(err)
+		case errors.Resemble(err, errRegistryOperation):
+			if errors.IsNotFound(errors.Cause(err)) {
+				return nil, interop.ErrUnknownDevEUI.WithCause(err)
+			}
+		}
+		return nil, interop.ErrJoinReq.WithCause(err)
+	}
+
+	header, err := req.AnswerHeader()
+	if err != nil {
+		return nil, interop.ErrMalformedMessage.WithCause(err)
+	}
+	ans := &interop.JoinAns{
+		JsNsMessageHeader: header,
+		PHYPayload:        interop.Buffer(res.RawPayload),
+		Result:            interop.ResultSuccess,
+		Lifetime:          uint32(res.Lifetime / time.Second),
+		AppSKey:           (*interop.KeyEnvelope)(res.AppSKey),
```

Comment:
Isn't there a nicer way to do this? A helper function perhaps?
This is not checked in any way at compile time AFAIK, so if any of types change, this will result in a panic at runtime.

Reply:
This is checked at compile time.

Also;

```
// KeyEnvelope contains a (encrypted) key.
type KeyEnvelope ttnpb.KeyEnvelope
```